components/causal_ts/src/tso.rs

```diff
@@ -62,55 +69,205 @@ impl TsoBatch {
+                Ok(_) => {
+                    return Some((
+                        TimeStamp::compose(self.physical, logical),
+                        logical + 1 == self.logical_end,
+                    ));
+                }
                 Err(x) => logical = x,
             }
         }
         None
     }
 
     // `last_ts` is the last timestamp of the new batch.
-    pub fn renew(&mut self, batch_size: u32, last_ts: TimeStamp) -> Result<()> {
-        let (physical, logical) = (last_ts.physical(), last_ts.logical() + 1);
-        let logical_start = logical.checked_sub(batch_size as u64).unwrap();
+    pub fn new(batch_size: u32, last_ts: TimeStamp) -> Self {
+        let (physical, logical_end) = (last_ts.physical(), last_ts.logical() + 1);
+        let logical_start = logical_end.checked_sub(batch_size as u64).unwrap();
+
+        Self {
+            size: batch_size,
+            physical,
+            logical_end,
+            logical_start: AtomicU64::new(logical_start),
+        }
+    }
+
+    /// Count of available TSO in the batch.
+    pub fn count(&self) -> u32 {
+        self.logical_end
+            .saturating_sub(self.logical_start.load(Ordering::Relaxed)) as u32
+    }
+
+    /// The original start timestamp in the batch.
+    pub fn original_start(&self) -> TimeStamp {
+        TimeStamp::compose(self.physical, self.logical_end - self.size as u64)
+    }
+
+    /// The excluded end timestamp after the last in batch.
+    pub fn excluded_end(&self) -> TimeStamp {
+        TimeStamp::compose(self.physical, self.logical_end)
+    }
+}
+
+/// `TsoBatchList` is a ordered list of `TsoBatch`. It aims to:
+/// 1. Cache more number of TSO to improve high availability. See issue #12794.
+///    `TsoBatch` can only cache at most 262144 TSO as logical clock is 18 bits.
+/// 2. Fully utilize cached TSO when some regions require latest TSO (e.g. in the scenario of leader transfer).
+///    Other regions without the requirement can still use older TSO cache.
+struct TsoBatchList {
+    inner: RwLock<TsoBatchListInner>,
+
+    /// Number of available TSO.
+    /// Using signed integer for avoiding a wrap around huge value as it's not precisely counted.
+    tso_count: AtomicI32,
+
+    /// Statistics of TSO usage.
+    tso_usage: AtomicU32,
+
+    /// Length of batch list. It is used to limit size for efficiency, and keep batches fresh.
+    capacity: u32,
+}
+
+// Inner data structure of batch list.
+// The reasons why `crossbeam_skiplist::SkipMap` is not chosen:
+// 1. In `flush()` procedure, a reader of `SkipMap` can still acquire a batch after the it is removed,
+//    which would violate the causality requirement. The `RwLock<BTreeMap>` avoid this scenario
+//    by lock synchronization.
+// 2. It is a scenario with much more reads than writes. The `RwLock` would be no less efficient
+//    than lock free implementation.
+type TsoBatchListInner = BTreeMap<u64, TsoBatch>;
```

Comment:
Why not choose LinkedList?  I think the timestamp in batch list is ordered and always pop the min ts and new allocated batch can push to back.

Reply:
To support the `after_ts` argument of `pop()`. Which is used in scenario (such as leader transfer) that some regions need latest TSO, but other regions can still use cached ones.